src/net/client/xfr.rs

```diff
@@ -0,0 +1,1147 @@
+/// emit an [`XfrEvent::TransferComplete`] event.
+///
+/// If the `TransferComplete` event has not been seen it means that the
+/// sequence is incomplete and the next response message in the sequence
+/// should be passed to [`process_next_answer()`] along with the exhausted
+/// iterator. This will populate thr [`XfrEventIterator`] with more records
+/// to parse thereby causing iteration to resume.
+///
+/// The process of producing and consuming iterators continues until the end
+/// of the transfer is detected or a parsing error occurs.
+pub struct XfrResponseProcessor;
+
+impl XfrResponseProcessor {
+    /// Process a single AXFR/IXFR response message.
+    ///
+    /// Return an [`XfrEventIterator`] over [`XfrEvent`]s emitted during
+    /// processing.
+    ///
+    /// If the returned iterator does not emit an
+    /// [`XfrEvent::TransferComplete`] event, call [`process_next_answer()`]
+    /// with the next response message to continue iterating over the transfer
+    /// responses.
+    pub async fn process_answer<'a>(
+        req: &'a Message<Bytes>,
+        resp: &'a Message<Bytes>,
+    ) -> Result<XfrEventIterator<'a>, Error> {
+        // Check that the given message is a DNS XFR response.
+        let xfr_type = Self::check_is_xfr_answer(req, resp).await?;
+
+        // https://datatracker.ietf.org/doc/html/rfc5936#section-3
+        // 3. Zone Contents "The objective of the AXFR session is to request
+        //   and transfer the contents of a zone, in order to permit the AXFR
+        //    client to faithfully reconstruct the zone as it exists at the
+        //    primary server for the given zone serial number.  The word
+        //    "exists" here designates the externally visible behavior, i.e.,
+        //    the zone content that is being served (handed out to clients) --
+        //    not its persistent representation in a zone file or database
+        //    used by the server -- and that for consistency should be served
+        //    subsequently by the AXFR client in an identical manner."
+        //
+        // So, walk over all the records in the answer, not just those that
+        // might be expected to exist in a zone (i.e. not just ZoneRecordData
+        // record types).
+        let answer = resp.answer().map_err(Error::ParseError)?;
+        let mut records = answer.into_records();
+
+        let Some(Ok(record)) = records.next() else {
+            return Err(Error::Malformed);
+        };
+
+        let state = Self::initialize(xfr_type, record).await?;
+
+        let records = records.flatten();
+
+        let iter = XfrEventIterator {
+            req,
+            state,
+            records,
+        };
+
+        Ok(iter)
+    }
+
+    /// Process a subsequent XFR response message.
+    ///
+    /// Revives the given iterator with new data based on the given response.
+    pub async fn process_next_answer<'a>(
+        it: &mut XfrEventIterator<'a>,
```

Comment:
Very confusing, an object with no state requires the caller to pass an iterator around that has the actual state.

Reply:
The iterator code has been reverted so this comment is outdated. Resolving.